### ISSUE TITLE
readiness checks: /server_info hangs when proxy is otherwise serving traffic.

### ISSUE THREAD

**Issue by mandarjog** [istio/istio#18046]
Readiness check depends on hitting `http://127.0.0.1:15000/server_info`

It gives rise to the following flipflops.
server_info is either serviced by the main thread, and when it is busy process the next config change it cannot respond to this message, or there is some lock involved somewhere.

We should use something else.

```
019-10-18T15:28:16.759902Z	info	Envoy proxy is NOT ready: config not received from Pilot (is Pilot running?): cds updates: 3 successful, 0 rejected; lds updates: 0 successful, 0 rejected
2019-10-18T15:28:17.509877Z	info	Envoy proxy is ready
2019-10-18T15:28:32.865884Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2019-10-18T15:28:33.583294Z	info	Envoy proxy is ready
2019-10-18T15:28:36.826089Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2019-10-18T15:28:37.515413Z	info	Envoy proxy is ready
2019-10-18T15:29:10.825813Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2019-10-18T15:29:10.889336Z	info	Envoy proxy is ready
```

**Comment by mandarjog** [comment]
@PiotrSikora @lambdai

**Comment by howardjohn** [comment]
Are you sure it can handle traffic fine in this case? I have seen this as well, but only when the pod is under heavy load. When I did test this, heavy load meant a bunch of XDS responses, so maybe this is blocked on the config path.

This is not actually flip flopping though, the readiness probe needs to fail 30 times in a row:
```
      readinessProbe:
        failureThreshold: 30
        httpGet:
          path: /healthz/ready
          port: 15020
          scheme: HTTP
        initialDelaySeconds: 1
        periodSeconds: 2
        successThreshold: 1
        timeoutSeconds: 1
```

**Comment by ramaraochavali** [comment]
Why do we have to check `/server_info` every time? Once it is live, should n't we catch the state and reuse it? It is not going to flipback to other state because Istio is not using active health checks
https://www.envoyproxy.io/docs/envoy/latest/api-v2/admin/v2alpha/server_info.proto#enum-admin-v2alpha-serverinfo-state

**Comment by ramaraochavali** [comment]
https://github.com/istio/istio/pull/18079 - PR to cache that state

**Comment by mandarjog** [comment]
@ramaraochavali does your PR resolve this issue?

**Comment by ramaraochavali** [comment]
@mandarjog Unfortunately that was reverted because there were concerns about hot restart. However I did another PR https://github.com/istio/istio/pull/18164 that simplifies how server_info is obtained which should help.

**Comment by nak3** [comment]
Hi @ramaraochavali @howardjohn

> I have seen this as well, but only when the pod is under heavy load. When I did test this, heavy load meant a bunch of XDS responses, so maybe this is blocked on the config path.

We are seeing very similar issue which @howardjohn mentioned above by Knative Serving's e2e test. The issue happened 100 pods scale tests, though does not happen 10 pods scale test. Also, 
strangely the issue started after upgraded t 1.3.3 from 1.3.2.

Should I open a separated ticket for that? Or perhaps https://github.com/istio/istio/pull/18164 will address it?

P.S. I am attaching the istio-proxy log for your reference [istio-proxy.log](https://github.com/istio/istio/files/3792505/istio-proxy.log)

**Comment by ramaraochavali** [comment]
@nak3 I looked at the log - your problem seems to be too much of initial config updates delaying proxy startup, IIUC your proxy is not even coming up. Is that correct?

If yes, you should try to trim your configuration using sidecar resource or some thing else. #18164 does not fully address your issue. So I suggest filing a different issue for discussion.

**Comment by ramaraochavali** [comment]
@howardjohn to your question "Are you sure it can handle traffic fine in this case" - Yes it would be able to handle because traffic is handled by worker threads. Admin endpoint processing and config ingestion from Pilot happen on main thread - so continuous updates/large updates from Pilot which takes more than a second to process in  Envoy can create these timeouts to admin endpoints. I handled that in #18164

**Comment by JRBANCEL** [comment]
> If yes, you should try to trim your configuration using sidecar resource or some thing else. #18164 does not fully address your issue. So I suggest filing a different issue for discussion.

This particular test was running fine before 1.3.3, so it is clearly a regression.
100 Knative Services translates to 200 VirtualServices. 100 Pods with sidecars are created.
Nothing extravagant for a 96 cores cluster with 300+GB of RAM.

**Comment by howardjohn** [comment]
Changes from 1.3.2 -> 1.3.3:

```
* 3cfabd9b3 "John Plevyak" <jplevyak@gmail.com> 3 weeks ago (tag: 1.3.3)
| Update envoy SHA for istio/istio#17699. (#2455) (#17780)
* f000a78aa "John Howard" <howardjohn@google.com> 3 weeks ago
| Read from metadata only as a fallback (#17766)
* a49c3ec2c "Martin Ostrowski" <mostrowski@google.com> 3 weeks ago
| Update operator SHA (#17749)
* 1d1fd0734 "Xinnan Wen" <iamwen@google.com> 3 weeks ago
| [1.3]Add perf benchmark script entry (#17642)
* 4ebeadb0f "Istio Automation" <istio.testing@gmail.com> 3 weeks ago
| Fix nil dereference from #17449 (#17458)
* 8b3320e5a "John Howard" <howardjohn@google.com> 3 weeks ago
| Fix 1.3 release test (#17710)
* 3c3a678cc "Romain Lenglet" <rlenglet@google.com> 3 weeks ago
| Update proxy SHA to HEAD of release-1.3 (#17696)
```

so if there was a regression, it appears it would have been in the proxy

**Comment by JRBANCEL** [comment]
I am not sure how this is possible given it doesn't seem to have changed, but the issue seems to be in `init_proxy`.
Running everything (proxyv2, pilot, ...) in `1.3.3` fails.
Running everything (proxyv2, pilot, ...) in `1.3.3`, but `init_proxy` in `1.3.2` works.

In 1.3.2, it is super fast: 
```
Normal   Pulled     2m51s                kubelet, gke-ksserving-e2e-cls-default-pool-092b0d29-7k62  Container image "docker.io/istio/proxy_init:1.3.2" already present on machine
Normal   Created    2m51s                kubelet, gke-ksserving-e2e-cls-default-pool-092b0d29-7k62  Created container istio-init
Normal   Started    2m51s                kubelet, gke-ksserving-e2e-cls-default-pool-092b0d29-7k62  Started container istio-init
Normal   Pulled     2m50s                kubelet, gke-ksserving-e2e-cls-default-pool-092b0d29-7k62  Container image "gcr.io/bancel-knative-dev/helloworld@sha256:65a146d2f451a28a10ff37963147b7615719ddd260922282d6ca98e2f7880634" already present on machine
```

In 1.3.3, it takes ~30s:
```
Normal   Pulled     2m12s              kubelet, gke-ksserving-e2e-cls-default-pool-092b0d29-7k62  Container image "docker.io/istio/proxy_init:1.3.3" already present on machine
Normal   Created    2m11s              kubelet, gke-ksserving-e2e-cls-default-pool-092b0d29-7k62  Created container istio-init
Normal   Started    118s               kubelet, gke-ksserving-e2e-cls-default-pool-092b0d29-7k62  Started container istio-init
Normal   Pulled     91s                kubelet, gke-ksserving-e2e-cls-default-pool-092b0d29-7k62  Container image "gcr.io/bancel-knative-dev/helloworld@sha256:65a146d2f451a28a10ff37963147b7615719ddd260922282d6ca98e2f7880634" already present on machine
```

Any idea?

**Comment by lambdai** [comment]
@JRBANCEL  IIUC istio is not involved between 
`Container image "docker.io/istio/proxy_init:1.3.3" already present on machine` and `Started container istio-init`. It's pure k8s work and the utilization of k8s node plays the role.
Is that particular node overloaded?
You could retry by schedule the pod to another node. Or simply delete the pod and retry.

**Comment by JRBANCEL** [comment]
I don't know. It is tricky to cleanly isolate specific pods.
All I know, is that creating 100 Pods with sidecar in the first scenario (all 1.3.3) results in some Pods not being ready after 3 minutes.
The exact same scenario in 1.3.2 has all Pods ready under 60s.
So something regressed somewhere.

**Comment by ramaraochavali** [comment]
@mandarjog for your timeouts issue, my PR got merged, can you please run your scenario and see if it helps?

**Comment by nak3** [comment]
Thank you @ramaraochavali @howardjohn @JRBANCEL @lambdai 

As @lambdai mentioned, I also suspected the CPU load and found an issue related Knative's specific structure so opened separated ticket https://github.com/istio/istio/issues/18632 (`Envoy proxy is NOT ready` was also reproduced.)

@mandarjog sorry for hijacking your report.

**Comment by vvbogdanov87** [comment]
I have the same problem with istio 1.4.2. Lots of messages in a sidecar log:
```
2020-01-09T08:00:23.233398Z     info    Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
```
As a result, lots of failed probes
```
Warning  Unhealthy  29m (x1070 over 5d23h)  kubelet, ip-*-*-*-* Readiness probe failed: Get http://*.*.*.*:15020/healthz/ready: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
```
And as a result, the pod was restarted 4 times in 6 days
```
NAME                           READY   STATUS    RESTARTS   AGE
***-75649fb797-k2j66   4/4     Running   4          6d
```

**Comment by vvbogdanov87** [comment]
This happens every 30 minutes.
```
[Envoy (Epoch 0)] [2020-01-09 23:07:38.954][16][warning][config] [bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:91] gRPC config stream closed: 13,
2020-01-09T23:30:09.233045Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-01-09T23:30:11.233568Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-01-09T23:30:13.233366Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-01-09T23:30:15.233421Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-01-09T23:30:16.863638Z	info	Envoy proxy is ready
[Envoy (Epoch 0)] [2020-01-09 23:35:54.555][16][warning][config] [bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:91] gRPC config stream closed: 13,
[Envoy (Epoch 0)] [2020-01-09 23:36:00.383][16][warning][config] [bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:91] gRPC config stream closed: 13,
2020-01-10T00:00:13.233179Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-01-10T00:00:13.998375Z	info	Envoy proxy is ready
[Envoy (Epoch 0)] [2020-01-10 00:07:34.261][16][warning][config] [bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:91] gRPC config stream closed: 13,
2020-01-10T00:30:07.233281Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-01-10T00:30:09.233153Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-01-10T00:30:11.233857Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-01-10T00:30:12.197421Z	info	Envoy proxy is ready
[Envoy (Epoch 0)] [2020-01-10 00:38:01.515][16][warning][config] [bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:91] gRPC config stream closed: 13,
[Envoy (Epoch 0)] [2020-01-10 00:50:03.147][16][warning][config] [bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:91] gRPC config stream closed: 13,
2020-01-10T01:00:07.233290Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-01-10T01:00:08.434695Z	info	Envoy proxy is ready
[Envoy (Epoch 0)] [2020-01-10 01:22:26.749][16][warning][config] [bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:91] gRPC config stream closed: 13,
[Envoy (Epoch 0)] [2020-01-10 01:28:31.206][16][warning][config] [bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:91] gRPC config stream closed: 13,
2020-01-10T01:30:07.234042Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-01-10T01:30:09.234011Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-01-10T01:30:10.599935Z	info	Envoy proxy is ready
2020-01-10T01:30:23.235385Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-01-10T01:30:23.730124Z	info	Envoy proxy is ready
[Envoy (Epoch 0)] [2020-01-10 01:35:01.502][16][warning][config] [bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:91] gRPC config stream closed: 13,
2020-01-10T02:00:07.233293Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-01-10T02:00:09.234513Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-01-10T02:00:11.233337Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-01-10T02:00:13.233082Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-01-10T02:00:15.233288Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-01-10T02:00:17.233348Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-01-10T02:00:19.233169Z	info	Envoy proxy is NOT ready: failed to get server info: failed retrieving Envoy stats: Get http://127.0.0.1:15000/server_info: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-01-10T02:00:20.601813Z	info	Envoy proxy is ready
```
I logged in to sidecar container and executed `curl http://127.0.0.1:15000/server_info` And it really doesn't respond during this periods. Looks like it is problem of envoy rather than istio.

**Comment by Siddharthk** [comment]
@howardjohn @ramaraochavali I am also facing this issue with Istio 1.6.7.  After 2-3 days of fresh deployment, the sidecar envoy proxy stops. Below are the logs: Note that the pod does not crash and is still in running state:
```
dev         app-amaikei6-6d6dc5c474-2sl84       1/2     Running            0          2d19h
```

Logs below:
```
2020-08-21T17:33:30.669986Z     info    Envoy proxy is ready
2020-08-21T18:01:58.066886Z     info    Subchannel Connectivity change to CONNECTING
```
After 2-3 days:
```
2020-08-24T08:16:09.602127Z     info    Channel Connectivity change to READY
2020-08-24T08:21:15.651523Z     warn    Envoy proxy is NOT ready: failed to get readiness stats: Get "http://127.0.0.1:15000/stats?usedonly&filter=^(server.state|listener_manager.workers_started)": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
2020-08-24T08:21:17.353903Z     info    Envoy proxy is ready
2020-08-24T08:21:27.694610Z     warn    Envoy proxy is NOT ready: failed to get readiness stats: Get "http://127.0.0.1:15000/stats?usedonly&filter=^(server.state|listener_manager.workers_started)": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
2020-08-24T08:21:29.790957Z     warn    Envoy proxy is NOT ready: failed to get readiness stats: Get "http://127.0.0.1:15000/stats?usedonly&filter=^(server.state|listener_manager.workers_started)": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
2020-08-24T08:21:31.823298Z     warn    Envoy proxy is NOT ready: failed to get readiness stats: Get "http://127.0.0.1:15000/stats?usedonly&filter=^(server.state|listener_manager.workers_started)": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
2020-08-24T08:21:33.989034Z     warn    Envoy proxy is NOT ready: failed to get readiness stats: Get "http://127.0.0.1:15000/stats?usedonly&filter=^(server.state|listener_manager.workers_started)": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
```

**Comment by mandarjog** [comment]
@Siddharthk do you happen to have cron jobs in your environment? At present we have an issue that cronjobs increase cardinality in an unbounded manner. 

`kubectl -n $ns -i -t exec $pod -c istio-proxy -- curl http://localhost:15000/stats | grep istio | wc -l`

**Comment by Siddharthk** [comment]
@mandarjog Actually its giving nothing:
```
[ec2-user@avd-4ece ~]$ kubectl -n dev -i -t exec app-amaikei6-6d6dc5c474-2sl84 -c istio-proxy -- curl -vvv http://localhost:15000/stats
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 15000 (#0)
> GET /stat0s HTTP/1.1
> Host: localhost:15000
> User-Agent: curl/7.58.0
> Accept: */*
>

```

**Comment by ceastman-ibm** [comment]
We are getting the same error with istio 1.6.8:
[istio-ingressgateway-684496d697-p882q] 2020-09-11T15:11:52.668870Z	warn	Envoy proxy is NOT ready: failed to get readiness stats: Get "http://127.0.0.1:15000/stats?usedonly&filter=^(server.state%7Clistener_manager.workers_started)": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

**Comment by howardjohn** [comment]
Fixed by #26820 which is now being backported
